File: Algorithmization/2semester/РГР алгоритмизация 2 сем.cs
using System;
using System.Collections.Generic;
using System.Text;

class Program
{
    static void Main(string[] args)
    {
        Console.OutputEncoding = Encoding.UTF8;
        MainMenu();
    }

    static void MainMenu()
    {
        while (true)
        {
            Console.WriteLine("\nМеню:");
            Console.WriteLine("1. Автор");
            Console.WriteLine("2. Посчитать правую польскую запись");
            Console.WriteLine("3. Проверка на правильность скобок");
            Console.WriteLine("4. Выход");

            Console.Write("Выберите пункт меню: ");
            string choice = Console.ReadLine();

            switch (choice)
            {
                case "1":
                    ShowAuthor();
                    break;
                case "2":
                    CalculateRPN();
                    break;
                case "3":
                    CheckBrackets();
                    break;
                case "4":
                    Console.WriteLine("Выход из программы.");
                    return;
                default:
                    Console.WriteLine("Некорректный выбор. Попробуйте еще раз.");
                    break;
            }
        }
    }

    static void ShowAuthor()
    {
        Console.WriteLine("\nАвтор: Бухинник Дмитрий Евгеньевич");
        Console.WriteLine("Нажмите Enter для возврата в меню.");
        Console.ReadLine();
    }

    static void CalculateRPN()
    {
        Console.Write("\nВведите правую польскую запись (через пробел): ");
        string expression = Console.ReadLine();
        string result = EvaluateRPN(expression);
        Console.WriteLine($"Результат: {result}");
        Console.WriteLine("Нажмите Enter для возврата в меню.");
        Console.ReadLine();
    }

    static string EvaluateRPN(string expression)
    {
        var ops = new Dictionary<string, Func<double, double, double>>
        {
            { "+", (a, b) => a + b },
            { "-", (a, b) => a - b },
            { "*", (a, b) => a * b },
            { "/", (a, b) => a / b }
        };

        var stack = new Stack<double>();
        foreach (var token in expression.Split())
        {
            if (ops.ContainsKey(token))
            {
                if (stack.Count < 2)
                    return "Ошибка: недостаточно операндов.";
                
                double b = stack.Pop();
                double a = stack.Pop();
                
                if (token == "/" && b == 0)
                    return "Ошибка: деление на ноль.";

                stack.Push(ops[token](a, b));
            }
            else
            {
                if (double.TryParse(token, out double number))
                {
                    stack.Push(number);
                }
                else
                {
                    return "Ошибка: некорректное значение.";
                }
            }
        }

        return stack.Count == 1 ? stack.Pop().ToString() : "Ошибка: некорректное выражение.";
    }

    static void CheckBrackets()
    {
        Console.Write("\nВведите выражение для проверки скобок: ");
        string expression = Console.ReadLine();
        bool result = IsBalanced(expression);
        Console.WriteLine(result ? "Скобки расставлены правильно." : "Скобки расставлены неправильно.");
        Console.WriteLine("Нажмите Enter для возврата в меню.");
        Console.ReadLine();
    }

    static bool IsBalanced(string expression)
    {
        var stack = new Stack<char>();
        var brackets = new Dictionary<char, char>
        {
            { '(', ')' },
            { '{', '}' },
            { '[', ']' }
        };

        foreach (var ch in expression)
        {
            if (brackets.ContainsKey(ch))
            {
                stack.Push(ch);
            }
            else if (brackets.ContainsValue(ch))
            {
                if (stack.Count == 0 || brackets[stack.Pop()] != ch)
                    return false;
            }
        }

        return stack.Count == 0;
    }
}

File: Algorithmization/РГР алгоритмизация 2 сем.py
def main_menu():
    while True:
        print("\nМеню:")
        print("1. Автор")
        print("2. Посчитать правую польскую запись")
        print("3. Проверка на правильность скобок")
        print("4. Выход")
        
        choice = input("Выберите пункт меню: ")
        
        if choice == '1':
            show_author()
        elif choice == '2':
            calculate_rpn()
        elif choice == '3':
            check_brackets()
        elif choice == '4':
            print("Выход из программы.")
            break
        else:
            print("Некорректный выбор. Попробуйте еще раз.")

def show_author():
    print("\nАвтор: Бухинник Дмитрий Евгеньевич")
    input("Нажмите Enter для возврата в меню.")

def calculate_rpn():
    import operator

    ops = {
        '+': operator.add,
        '-': operator.sub,
        '*': operator.mul,
        '/': operator.truediv,
    }

    def evaluate_rpn(expression):
        stack = []
        for token in expression.split():
            if token in ops:
                if len(stack) < 2:
                    return "Ошибка: недостаточно операндов."
                b, a = stack.pop(), stack.pop()
                if token == '/' and b == 0:
                    return "Ошибка: деление на ноль."
                stack.append(ops[token](a, b))
            else:
                try:
                    stack.append(float(token))
                except ValueError:
                    return "Ошибка: некорректное значение."
        if len(stack) != 1:
            return "Ошибка: некорректное выражение."
        return stack[0]

    expression = input("\nВведите правую польскую запись (через пробел): ")
    result = evaluate_rpn(expression)
    print(f"Результат: {result}")
    input("Нажмите Enter для возврата в меню.")

def check_brackets():
    def is_balanced(expression):
        stack = []
        brackets = {'(': ')', '{': '}', '[': ']'}
        for char in expression:
            if char in brackets:
                stack.append(char)
            elif char in brackets.values():
                if not stack or brackets[stack.pop()] != char:
                    return False
        return not stack

    expression = input("\nВведите выражение для проверки скобок: ")
    if is_balanced(expression):
        print("Скобки расставлены правильно.")
    else:
        print("Скобки расставлены неправильно.")
    input("Нажмите Enter для возврата в меню.")

if __name__ == "__main__":
    main_menu()
